Source/WebKit/WebProcess/Extensions/Cocoa/_WKWebExtensionWebRequestFilter.mm

```diff
@@ -0,0 +1,248 @@
+    NSSet<NSNumber *> *_types;
+}
+
+- (instancetype)initWithDictionary:(NSDictionary<NSString *, id> *)dictionary outErrorMessage:(NSString **)outErrorMessage
+{
+    if (!(self = [super init])) {
+        *outErrorMessage = @"Runtime failure.";
+        return nil;
+    }
+
+    static NSArray<NSString *> *requiredKeys = @[
+        urlsKey,
+    ];
+
+    static NSDictionary<NSString *, id> *expectedTypes = @{
+        urlsKey: @[ NSString.class ],
+        typesKey: NSArray.class,
+        tabIdKey: NSNumber.class,
+        windowIdKey: NSNumber.class,
+    };
+
+    if (!validateDictionary(dictionary, nil, requiredKeys, expectedTypes, outErrorMessage))
+        return nil;
+
+    if ((*outErrorMessage = [self _initializeWithDictionary:dictionary]))
+        return nil;
+
+    return self;
+}
+
+static NSArray<_WKWebExtensionMatchPattern *> *toURLs(NSArray<NSString *> *value, NSString **outErrorMessage)
+{
+    if (!value.count)
+        return nil;
+
+    NSError *error;
+
+    NSMutableArray<_WKWebExtensionMatchPattern *> *patterns = [[NSMutableArray alloc] init];
+    for (NSString *rawPattern in value) {
+        _WKWebExtensionMatchPattern *pattern = [[_WKWebExtensionMatchPattern alloc] initWithString:rawPattern error:&error];
+        if (!pattern) {
+            if (outErrorMessage)
+                *outErrorMessage = error.userInfo[NSDebugDescriptionErrorKey];
+            return nil;
+        }
+        [patterns addObject:pattern];
+    }
+
+    return patterns;
+}
+
+static NSNumber *toResourceType(NSString *typeString, NSString **outErrorMessage)
+{
+    static NSDictionary<NSString *, NSNumber *> *validTypes = @{
+        @"main_frame": @(_WKWebExtensionWebRequestResourceTypeMainFrame),
+        @"sub_frame": @(_WKWebExtensionWebRequestResourceTypeSubframe),
+        @"stylesheet": @(_WKWebExtensionWebRequestResourceTypeStylesheet),
+        @"script": @(_WKWebExtensionWebRequestResourceTypeScript),
+        @"image": @(_WKWebExtensionWebRequestResourceTypeImage),
+        @"font": @(_WKWebExtensionWebRequestResourceTypeFont),
+        @"object": @(_WKWebExtensionWebRequestResourceTypeObject),
+        @"xmlhttprequest": @(_WKWebExtensionWebRequestResourceTypeXMLHTTPRequest),
+        @"ping": @(_WKWebExtensionWebRequestResourceTypePing),
+        @"csp_report": @(_WKWebExtensionWebRequestResourceTypeCSPReport),
+        @"media": @(_WKWebExtensionWebRequestResourceTypeMedia),
+        @"websocket": @(_WKWebExtensionWebRequestResourceTypeWebsocket),
+        @"web_manifest": @(_WKWebExtensionWebRequestResourceTypeApplicationManifest),
+        @"xslt": @(_WKWebExtensionWebRequestResourceTypeXSLT),
+        @"beacon": @(_WKWebExtensionWebRequestResourceTypeBeacon),
+        @"other": @(_WKWebExtensionWebRequestResourceTypeOther),
+    };
+
+    NSNumber *typeAsNumber = validTypes[typeString];
+    if (!typeAsNumber) {
+        *outErrorMessage = @"'types' must be an array of ResourceType.";
```

Comment:
Should use `toErrorString(nil, typesKey, @"'%@' is an unknown resource type", typeString)`.